File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.java
import java.util.*;
import java.io.*;

class Solution {
    
    // 북 / 동 / 남 / 서
    static int[] dx = {-1,0,1,0};
    static int[] dy = {0,1,0,-1};
    
    static int N,M;
    static int[][] visited;
    static int[] oil;
    
    public int solution(int[][] land) {
        int answer = 0;
        
        N = land.length;
        M = land[0].length;
        
        visited = new int[N][M];
        oil = new int[M];
        
        for (int i=0; i < N; i++) {
            for (int j=0; j < M; j++) {
                if (land[i][j] == 1 && visited[i][j] == 0) {
                    bfs(i,j,land);
                }
            }
        }
        for (int value : oil) {
            answer = Math.max(answer,value);
        }
        
        return answer;
    }
    
    public void bfs(int x, int y, int[][] land) {
        Queue<int[]> q = new LinkedList<>();
        Set<Integer> set = new HashSet<>();
        int cnt = 1;
        
        visited[x][y] = 1;
        q.add(new int[]{x,y});
        
        while (!q.isEmpty()) {
            int[] s = q.poll();
            int sx = s[0];
            int sy = s[1];
            set.add(sy);
            
            for (int d=0; d < 4; d++) {
                int nx = sx + dx[d];
                int ny = sy + dy[d];
                
                if (nx >= 0 && nx < N && ny >= 0 && ny < M) {
                    if (visited[nx][ny] == 0 && land[nx][ny] == 1) {
                        visited[nx][ny] = 1;
                        cnt += 1;
                        q.add(new int[]{nx,ny});                        
                    }
                }
            }    
        }
        
        for (int v : set) {
            oil[v] += cnt;
        }
    }
}

File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.py
from collections import deque

def solution(land):
    answer = 0
    
    # 4방향 북/동/남/서
    dx = [-1,0,1,0]
    dy = [0,1,0,-1]
    
    visited = [[0 for j in range(len(land[0]))] for i in range(len(land))]
    answer = [0 for j in range(len(land[0]))]
    
    def bfs(a,b) :
        q = deque()
        q.append((a,b))
        cnt = 0
        temp = []
        visited[a][b] = 1
        
        while q :
            x,y = q.popleft()
            cnt += 1
            temp.append(y)
            
            for i in range(4) :
                nx = x + dx[i]
                ny = y + dy[i]
                
                if 0 <= nx < len(land) and 0 <= ny < len(land[0]) and visited[nx][ny] == 0 and land[nx][ny] == 1 :
                    q.append((nx,ny))
                    visited[nx][ny] = 1
        y_min = min(temp)
        y_max = max(temp)
        ## print(temp,cnt)
        for i in range(y_min,y_max+1) :
            answer[i] += cnt
            
    for i in range(len(land)) :
        for j in range(len(land[0])) :
            if land[i][j] == 1 and visited[i][j] == 0:
                bfs(i,j)
                
    return max(answer)